test/test_torch.py

```diff
@@ -17228,6 +17228,122 @@ def test_reciprocal_complex_extremal(self, device, dtype):
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([-1.0, -0.0, -0.0, -1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, -0.0)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([1.0, 0.0, 0.0, 1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, 0.0)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        # Normal
+        expected = torch.tensor([[1.0, 0.0, 0.0, 1.0] * 2, [-1.0, -0.0, -0.0, -1.0] * 2], dtype=dtype).reshape(-1)
+        x = torch.tensor([-1.0, -0.0, 0.0, 1.0] * 4, dtype=dtype)
+        y = torch.tensor([[0.0] * 4, [1.0] * 4, [-0.0] * 4, [-1.0] * 4], dtype=dtype).reshape(-1)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+    @dtypes(torch.int8, torch.short, torch.int, torch.long)
+    def test_copysign_integral(self, device, dtype):
+        # integer -0 doesn't exist
+        # Result is promoted to float
+        x = torch.tensor([-1, 0, 1], dtype=dtype)
+        y = torch.tensor(-1)
+        expected = torch.tensor([-1.0, -0.0, -1.0], dtype=torch.float)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1, 0, 1], dtype=dtype)
+        y = torch.tensor(1)
+        expected = torch.tensor([1.0, 0.0, 1.0], dtype=torch.float)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+    def _test_copysign_numpy(self, a, b):
+        torch_result = torch.copysign(a, b)
+        expected = torch.from_numpy(np.copysign(a.cpu().numpy(), b.cpu().numpy()))
+
+        # To handle inconsistencies in our type promotion and numpy's type promotion
+        promoted_type = torch.promote_types(torch_result.dtype, expected.dtype)
+        torch_result = torch_result.to(promoted_type)
+        expected = expected.to(promoted_type)
+
+        self.assertEqual(torch_result, expected)
+
+    def _create_rand_tensor(self, size, dtype, device):
+        if dtype == torch.bool:
+            return torch.randint(high=2, size=size, dtype=dtype, device=device)
+        elif dtype in [torch.int8, torch.short, torch.int, torch.long]:
+            return torch.randint(high=10, size=size, dtype=dtype, device=device)
+        else:
+            return torch.randn(size, dtype=dtype, device=device)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_copysign_numpy(self, device):
```

Comment:
Use the @dtypes decorator and the device-specific versions to test half, bfloat16, float and double. For bfloat16 you can compare with NumPy by running the NumPy values in float and then converting back to bfloat16. The dtypes decorators can accept a cross product of dtypes so you don't need to enumerate them in the test.